pypa/warehouse
# Show “API Endpoint” error page for invalid requests to upload.pypi.org/legacy/

Addresses https://github.com/pypa/warehouse/issues/2288. Maybe also addresses #2225. From the issue:

> URLs like https://upload.pypi.org/legacy/ and https://test.pypi.org/legacy/ currently return a 404 error when accessed via GET. [...] It would be great if PyPI returned a more informative response when accessing these URLs.

This change replaces the 404 response mentioned in the issue with the "API Endpoint" HTML error page already in use the https://upload.pypi.org/ URL. The change applies to any request to https://upload.pypi.org/legacy/ that is not handled by another route already.


### Possible discussion points

* This changes the response code for the affected URLs from 404 to 200. Non web browser clients might be confused by this.
* Ideally, all URLs that return the "API Endpoint" template should probably come with a status code `400 Bad Request`. 
* Test coverage: I couldn't figure out how to write a test for this. Maybe someone can point me existing tests that cover routing and the special forklift domain?


### How was this tested?

To test and reproduce this, I took the following steps:

1. Add two domains to `/etc/hosts` for realistic testing of upload.pypi.org vs pypi.org routes:
   ```
   127.0.0.1   pypi.local
   127.0.0.1   upload.pypi.local
   ```
2. Set two environment variables for `app` and derived containers in `docker-compose.yaml`:
   ```yaml
   app:
     # ...
     environment:
       # ...
       FORKLIFT_DOMAIN: upload.pypi.local
       WAREHOUSE_DOMAIN: pypi.local
   ```
3. Visit these URLs in a webbrowser and confirm that I see what's in this screenshot (new behavior):
   * http://upload.pypi.local/
   * http://upload.pypi.local/legacy
   * http://upload.pypi.local/legacy/
   * http://upload.pypi.local/legacy/?:action=file_upload2
   * http://upload.pypi.local/legacy/?:schadenfreude
   <img width="1280" alt="screen shot 2017-09-17 at 1 06 07 pm" src="https://user-images.githubusercontent.com/97921/30524482-acfca136-9ba9-11e7-80f5-cd167f5ee88a.png">
4. Visit these URLs in a web browser and confirm that I see this screenshot (unchanged behavior):
   * http://upload.pypi.local/fernweh
   * http://upload.pypi.local/wiener/schnitzel
   <img width="1280" alt="screen shot 2017-09-17 at 1 06 19 pm" src="https://user-images.githubusercontent.com/97921/30524483-b19117d6-9ba9-11e7-9b21-ad123a054c78.png">
5. Visit this URL in a web browser and confirm that I see this screenshot (unchanged behavior):
   * http://upload.pypi.local/legacy/?:action=file_upload
   <img width="1280" alt="screen shot 2017-09-17 at 1 08 10 pm" src="https://user-images.githubusercontent.com/97921/30524489-bb9abd18-9ba9-11e7-801d-b306cf69d491.png">
